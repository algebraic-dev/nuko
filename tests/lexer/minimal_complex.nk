type Range = { line: Int }

type Hole =
  | Filled Range
  | Empty

-- Without lambda match
let isEmpty? (hole: Hole): Bool =
  match hole with
    Filled range => False
    Empty        => True

-- Lambda Case version of isEmpty?
let isEmpty? : Hole -> Bool = match
  Filled range => False
  Empty        => True

let isEmpty! (hole: Hole): Unit =
  if isEmpty? hole
    then print "It's empty!!"
    else print "Is's not empty.'"

let getLineOr (hole: Hole) (alt: Int): Int =
  match hole with
    Filled range => range.line
    Empty        => alt

let sequenced!: Unit =
  print "a"
  print "b"
  print "c"