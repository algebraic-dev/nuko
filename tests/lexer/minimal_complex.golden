(type,1:0-1:4)
(upper(Range),1:5-1:10)
('=',1:11-1:12)
('{',1:13-1:14)
(lower(line),1:15-1:19)
(':',1:19-1:20)
(upper(Int),1:21-1:24)
('}',1:25-1:26)
(type,3:0-3:4)
(upper(Hole),3:5-3:9)
('=',3:10-3:11)
('|',4:2-4:3)
(upper(Filled),4:4-4:10)
(upper(Range),4:11-4:16)
('|',5:2-5:3)
(upper(Empty),5:4-5:9)
(let,8:0-8:3)
(lower(isEmpty?),8:4-8:12)
('(',8:13-8:14)
(lower(hole),8:14-8:18)
(':',8:18-8:19)
(upper(Hole),8:20-8:24)
(')',8:24-8:25)
(':',8:25-8:26)
(upper(Bool),8:27-8:31)
('=',8:32-8:33)
($begin,9:2-9:2)
(match,9:2-9:7)
(lower(hole),9:8-9:12)
(with,9:13-9:17)
($begin,10:4-10:4)
(upper(Filled),10:4-10:10)
(lower(range),10:11-10:16)
(=>,10:17-10:19)
($begin,10:20-10:20)
(upper(False),10:20-10:25)
($end,11:4-11:4)
($semi,11:4-11:4)
(upper(Empty),11:4-11:9)
(=>,11:17-11:19)
($begin,11:20-11:20)
(upper(True),11:20-11:24)
($end,13:0-13:0)
($end,13:0-13:0)
($end,13:0-13:0)
(let,14:0-14:3)
(lower(isEmpty?),14:4-14:12)
(':',14:13-14:14)
(upper(Hole),14:15-14:19)
(->,14:20-14:22)
(upper(Bool),14:23-14:27)
('=',14:28-14:29)
($begin,14:30-14:30)
(match,14:30-14:35)
($end,15:2-15:2)
(upper(Filled),15:2-15:8)
(lower(range),15:9-15:14)
(=>,15:15-15:17)
($begin,15:18-15:18)
(upper(False),15:18-15:23)
($end,16:2-16:2)
(upper(Empty),16:2-16:7)
(=>,16:15-16:17)
($begin,16:18-16:18)
(upper(True),16:18-16:22)
($end,18:0-18:0)
(let,18:0-18:3)
(lower(isEmpty!),18:4-18:12)
('(',18:13-18:14)
(lower(hole),18:14-18:18)
(':',18:18-18:19)
(upper(Hole),18:20-18:24)
(')',18:24-18:25)
(':',18:25-18:26)
(upper(Unit),18:27-18:31)
('=',18:32-18:33)
($begin,19:2-19:2)
(if,19:2-19:4)
($begin,19:5-19:5)
(lower(isEmpty?),19:5-19:13)
(lower(hole),19:14-19:18)
($end,20:4-20:4)
(then,20:4-20:8)
($begin,20:9-20:9)
(lower(print),20:9-20:14)
("It's empty!!",20:15-20:29)
($end,21:4-21:4)
(else,21:4-21:8)
($begin,21:9-21:9)
(lower(print),21:9-21:14)
("Is's not empty.'",21:15-21:33)
($end,23:0-23:0)
($end,23:0-23:0)
(let,23:0-23:3)
(lower(getLineOr),23:4-23:13)
('(',23:14-23:15)
(lower(hole),23:15-23:19)
(':',23:19-23:20)
(upper(Hole),23:21-23:25)
(')',23:25-23:26)
('(',23:27-23:28)
(lower(alt),23:28-23:31)
(':',23:31-23:32)
(upper(Int),23:33-23:36)
(')',23:36-23:37)
(':',23:37-23:38)
(upper(Int),23:39-23:42)
('=',23:43-23:44)
($begin,24:2-24:2)
(match,24:2-24:7)
(lower(hole),24:8-24:12)
(with,24:13-24:17)
($begin,25:4-25:4)
(upper(Filled),25:4-25:10)
(lower(range),25:11-25:16)
(=>,25:17-25:19)
($begin,25:20-25:20)
(lower(range),25:20-25:25)
('.',25:25-25:26)
(lower(line),25:26-25:30)
($end,26:4-26:4)
($semi,26:4-26:4)
(upper(Empty),26:4-26:9)
(=>,26:17-26:19)
($begin,26:20-26:20)
(lower(alt),26:20-26:23)
($end,28:0-28:0)
($end,28:0-28:0)
($end,28:0-28:0)
(let,28:0-28:3)
(lower(sequenced!),28:4-28:14)
(':',28:14-28:15)
(upper(Unit),28:16-28:20)
('=',28:21-28:22)
($begin,29:2-29:2)
(lower(print),29:2-29:7)
("a",29:8-29:11)
($semi,30:2-30:2)
(lower(print),30:2-30:7)
("b",30:8-30:11)
($semi,31:2-31:2)
(lower(print),31:2-31:7)
("c",31:8-31:11)
($end,31:11-31:11)
($EOF,31:11-31:11)
